retworkx-core/src/centrality.rs

```diff
@@ -297,3 +300,201 @@ where
+/// [`eigenvector_centrality()`](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.centrality.eigenvector_centrality.html)
+/// function.
+///
+/// In the case of multigraphs the weights of any parallel edges will be
+/// summed when computing the eigenvector centrality.
+///
+/// Arguments:
+///
+/// * `graph` - The graph object to run the algorithm on
+/// * `weight_fn` - An input callable that will be pased the `EdgeRef` for
+///     an edge in the graph and is expected to return a `Result<f64>` of
+///     the weight of that edge.
+/// * `max_iter` - The maximum number of iterations in the power method. If
+///     set to `None` a default value of 100 is used.
+/// * `tol` - The error tolerance used when checking for convergence in the
+///     power method. If set to `None` a dfault value of 1e-6 is used.
+///
+/// # Example
+/// ```rust
+/// use hashbrown::HashMap;
+///
+/// use retworkx_core::Result;
+/// use retworkx_core::petgraph;
+/// use retworkx_core::petgraph::visit::{IntoEdges, IntoNodeIdentifiers};
+/// use retworkx_core::centrality::eigenvector_centrality;
+///
+/// let g = petgraph::graph::UnGraph::<i32, ()>::from_edges(&[
+///     (0, 1), (1, 2)
+/// ]);
+/// // Calculate the betweeness centrality
+/// let output: Result<Option<HashMap<petgraph::graph::NodeIndex, f64>>> =
+///     eigenvector_centrality(&g, |_| {Ok(1.)}, None, None);
+/// ```
+pub fn eigenvector_centrality<G, F, E>(
+    graph: G,
+    mut weight_fn: F,
+    max_iter: Option<usize>,
+    tol: Option<f64>,
+) -> Result<Option<HashMap<G::NodeId, f64>>, E>
+where
+    G: NodeIndexable
+        + IntoNodeIdentifiers
+        + IntoNeighbors
+        + IntoEdges
+        + NodeCount
+        + GraphProp
+        + GraphBase<NodeId = NodeIndex>,
+    F: FnMut(G::EdgeRef) -> Result<f64, E>,
+{
+    let tol: f64 = tol.unwrap_or(1e-6);
+    let max_iter = max_iter.unwrap_or(50);
+    let n_start: HashMap<G::NodeId, f64> = graph.node_identifiers().map(|n| (n, 1.)).collect();
+    let n_start_sum: f64 = n_start.len() as f64;
+    let mut x: HashMap<G::NodeId, f64> =
+        n_start.iter().map(|(k, v)| (*k, v / n_start_sum)).collect();
+    let node_count = graph.node_count();
+    for _ in 0..max_iter {
+        let x_last = x.clone();
+        for node in x_last.keys() {
+            for neighbor in graph.neighbors(*node) {
+                let w_vec: Vec<G::EdgeRef> = graph
+                    .edges(*node)
+                    .filter(|edge| edge.target() == neighbor)
+                    .collect();
+                let mut w = 0.;
+                for edge in w_vec {
+                    w += weight_fn(edge)?;
+                }
+                *x.get_mut(&neighbor).unwrap() += x_last[node] * w;
+            }
+        }
+        let mut norm: f64 = x.values().map(|val| val.powi(2)).sum::<f64>().sqrt();
+        if norm == 0. {
+            norm = 1.;
+        }
```

Comment:
If this ever happens it'd mean our current vector is very close to all-zero and a subsequent iteration might not be able to change it much so we'll think that we converged. But convergence to an all-zero solution is not a right solution so I'm thinking we might want to return `None` here.

Reply:
That's fine with me, I'll update this: 
```suggestion
        if norm == 0. {
            return Ok(None);
        }
```